typescript/sdk/src/gcp/validator.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  Announcement,
+  BaseValidator,
+  S3Announcement,
+  S3CheckpointWithId,
+  ValidatorConfig,
+  ValidatorMetadata,
+  isS3CheckpointWithId,
+} from '@hyperlane-xyz/utils';
+
+import { GcpStorageWrapper, StorageConfig } from './storage.js';
+
+const checkpointWithMessageIdKey = (checkpointIndex: number) =>
+  `checkpoint_${checkpointIndex}_with_id.json`;
+const LATEST_KEY = 'gcsLatestIndexKey';
+const ANNOUNCEMENT_KEY = 'gcsAnnouncementKey';
+const METADATA_KEY = 'gcsMetadataKey';
+const LOCATION_PREFIX = 'gs://';
+
+/**
+ * Extension of BaseValidator that includes GCP Cloud Storage utilities.
+ */
+export class GcpValidator extends BaseValidator {
+  public storage: GcpStorageWrapper;
+
+  constructor(
+    public validatorConfig: ValidatorConfig,
+    public storageConfig: StorageConfig,
+  ) {
+    super(validatorConfig);
+    this.storage = new GcpStorageWrapper(storageConfig);
+  }
+
+  static async fromStorageLocation(
+    storageLocation: string,
+  ): Promise<GcpValidator> {
+    // Remove trailing key if present
+    if (storageLocation.endsWith(ANNOUNCEMENT_KEY)) {
+      storageLocation = storageLocation.slice(0, -ANNOUNCEMENT_KEY.length);
+      // Remove trailing slash if present after removing key
+      if (storageLocation.endsWith('/')) {
+        storageLocation = storageLocation.slice(0, -1);
+      }
+    }
+    if (storageLocation.startsWith(LOCATION_PREFIX)) {
+      const bucketName = storageLocation.slice(LOCATION_PREFIX.length);
+      const pieces = bucketName.split('/');
+      if (pieces.length >= 1) {
+        const storageConfig = {
+          bucket: pieces[0],
+          folder: pieces.slice(1).join('/'),
```

Comment:
iiuc this will set `folder` to `''` instead of `undefined` if there's no folder, is that intentional?